[Enhancement] Support Async delete files from Recycle Bin

srlch opened the pull request:
## Why I'm doing:
Currently, delete data in a sync operation for shared data mode in CatalogBin. The delete operation may take a very long time with the object lock for CatalogBin which may stuck sometime.

## What I'm doing:
Make file deletion in async way for both shared data mode and shared nothing mode.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

kevincai commented on fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:
```diff
@@ -494,15 +508,33 @@ void eraseTable(long currentTimeMs) {
+
+                if (!finished) {
+                    // finish with error, re-submit in next round
+                    asyncDeleteForTables.remove(info);
```
worth logging with the error, in case of endless retry.

srlch replied:
fixed

srlch committed:
[Enhancement] Support Async delete files from Recycle Bin in shared data mode

Why I'm doing:
Currently, delete data in a sync operation for shared data mode in CatalogBin.
The delete operation may take a very long time with the object lock for CatalogBin
which may stuck sometime.

What I'm doing:
Make file deletion in async way for shared data mode.

Signed-off-by: srlch <linzichao@starrocks.com>

srlch commented on fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java:
```diff
@@ -553,7 +605,7 @@ protected synchronized void erasePartition(long currentTimeMs) {
                     break;
                 }
-            } else {
+            } else if (asyncDeleteForPartitions.get(partitionInfo) == null) /* treated as error if task is not running */ {
```
fixed

srlch commented on fe/fe-core/src/main/java/com/starrocks/lake/Utils.java:
```diff
@@ -221,4 +222,29 @@ public static Optional<Long> getWarehouseIdByNodeId(SystemInfoService systemInfo
+    public static boolean checkAllAsyncTaskFinished(List<CompletableFuture<Boolean>> asyncTaskReturn) {
+        boolean allFinished = false;
+        allFinished = asyncTaskReturn.stream().allMatch(future -> future != null && future.isDone());
+        return allFinished;
```
remove this useless code.

srlch committed:
fix ut

Signed-off-by: srlch <linzichao@starrocks.com>
